File: Core/Catalogues/DatabaseCatalogue.cs
using KolibSoft.Catalogue.Core.Abstractions;
using KolibSoft.Catalogue.Core.Utils;
using Microsoft.EntityFrameworkCore;
using static KolibSoft.Catalogue.Core.Constants;

namespace KolibSoft.Catalogue.Core.Catalogues;

public class DatabaseCatalogue<TItem, TFilters> : ICatalogueConnector<TItem, TFilters>
    where TItem : class, IItem, IValidatable, IUpdatable<TItem>
{

    public DbContext DbContext { get; }
    public DbSet<TItem> DbSet { get; }
    public ICollection<string> Errors { get; } = new List<string>();
    public virtual bool Available => DbContext.Database.CanConnect();

    protected virtual Task<IQueryable<TItem>> OnQueryAsync(TFilters? filters) => Task.FromResult<IQueryable<TItem>>(DbSet);
    protected virtual Task<TItem?> OnGetAsync(Guid id) => Task.Run(() => DbSet.FirstOrDefault(x => x.Id == id));
    protected virtual Task<bool> OnValidateInsertAsync(TItem item) => Task.FromResult(true);
    protected virtual Task<bool> OnValidateUpdateAsync(TItem item) => Task.FromResult(true);
    protected virtual Task<bool> OnValidateDeleteAsync(TItem item) => Task.FromResult(true);

    public virtual async Task<Result<Page<TItem>?>> QueryAsync(TFilters? filters = default, int pageIndex = 0, int pageSize = DefaultChunkSize)
    {
        Errors.Clear();
        var items = await OnQueryAsync(filters);
        var page = items.GetPage(pageIndex, pageSize);
        return page;
    }

    public virtual async Task<Result<TItem?>> GetAsync(Guid id)
    {
        Errors.Clear();
        var item = await OnGetAsync(id);
        return item;
    }

    public virtual async Task<Result<TItem?>> InsertAsync(TItem item)
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        var original = DbSet.FirstOrDefault(x => x.Id == item.Id);
        if (original != null) return default(TItem?);
        if (!await OnValidateInsertAsync(item)) return Errors.ToArray();
        DbSet.Add(item);
        DbContext.SaveChanges();
        return item;
    }

    public virtual async Task<Result<TItem?>> UpdateAsync(Guid id, TItem item)
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        var original = DbSet.FirstOrDefault(x => x.Id == id);
        if (original == null) return default(TItem?);
        if (!await OnValidateUpdateAsync(item)) return Errors.ToArray();
        if (original.ModifiedAt <= item.ModifiedAt) original.Update(item);
        DbSet.Update(item);
        DbContext.SaveChanges();
        return original;
    }

    public virtual async Task<Result<TItem?>> DeleteAsync(Guid id)
    {
        Errors.Clear();
        var item = DbSet.FirstOrDefault(x => x.Id == id);
        if (item == null) return default(TItem?);
        if (!await OnValidateDeleteAsync(item)) return Errors.ToArray();
        DbSet.Remove(item);
        DbContext.SaveChanges();
        return item;
    }

    public DatabaseCatalogue(DbContext dbContext)
    {
        DbContext = dbContext;
        DbSet = DbContext.Set<TItem>();
    }

}

File: docs/lib/catalogues/database_catalogue.js
import { Page } from "../page.js";
import { Result } from "../result.js";
import { PageUtils } from "../utils/page_utils.js";
import { DbContext } from "../database/dbcontext.js";
import { DbSet } from "../database/dbset.js";
import { DefaultChunkSize } from "../constants.js";
import { CatalogueConnector } from "../abstractions/catalogue_connector.js";

/**
 * @template TItem
 * @template TFilters
 * @implements {CatalogueConnector<TItem, TFilters>}
 */
class DatabaseCatalogue extends CatalogueConnector {

    /** @type {(json: {}) => TItem} */
    #creator;

    /** @type {DbContext} */
    #dbContext;

    /** @type {DbSet<TItem>} */
    #dbSet;

    /** @type {string[]} */
    #errors;

    /** @returns {(json: {}) => TItem} */
    get creator() { return this.#creator; }

    /** @returns {DbContext} */
    get dbContext() { return this.#dbContext; }

    /** @returns {DbSet<TItem>} */
    get dbSet() { return this.#dbSet; }

    /** @returns {string[]} */
    get errors() { return this.#errors; }

    /** @returns {boolean} */
    get available() { return 'indexedDB' in window; }

    /**
     * @param {TFilters?} filters 
     * @returns {TItem[]}
     */
    async onQueryAsync(filters) { return await this.#dbSet.filterAsync(x => true); }

    /**
     * @param {string} id 
     * @returns {TItem?}
     */
    async onGetAsync(id) { return await this.#dbSet.getAsync(id); }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    async onValidateInsertAsync(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    async onValidateUpdateAsync(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    async onValidateDeleteAsync(item) { return true; }

    /**
     * @param {TFilters?} filters 
     * @param {number} pageIndex 
     * @param {number} pageSize 
     * @returns {Promise<Result<Page<TItem>?>>}
     */
    async queryAsync(filters = null, pageIndex = 0, pageSize = DefaultChunkSize) {
        this.#errors = [];
        let items = await this.onQueryAsync(filters);
        if (this.#creator) items = items.map(x => this.#creator(x));
        let page = PageUtils.getPage(items, pageIndex, pageSize);
        return new Result({ data: page });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem?>>}
     */
    async getAsync(id) {
        this.#errors = [];
        let item = await this.onGetAsync(id);
        if (this.#creator && item != null) item = this.#creator(item);
        return new Result({ data: item });
    }

    /**
     * @param {TItem} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async insertAsync(item) {
        this.#errors = [];
        if (!item.validate(this.#errors)) return new Result({ errors: this.#errors });
        let original = await this.#dbSet.getAsync(item.id);
        if (original != null) return new Result({ data: null });
        if (!(await this.onValidateInsertAsync(item))) return new Result({ errors: this.#errors });
        await this.#dbSet.addAsync(item);
        return new Result({ data: item });
    }

    /**
     * @param {string} id 
     * @param {TItem} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async updateAsync(id, item) {
        this.#errors = [];
        if (!item.validate(this.#errors)) return new Result({ errors: this.#errors });
        let original = await this.#dbSet.getAsync(id);
        if (original == null) return new Result({ data: null });
        if (this.#creator) original = this.#creator(original);
        if (!(await this.onValidateUpdateAsync(item))) return new Result({ errors: this.#errors });
        original.update(item);
        await this.#dbSet.updateAsync(item);
        return new Result({ data: original });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem?>>}
     */
    async deleteAsync(id) {
        this.#errors = [];
        let item = await this.#dbSet.getAsync(id)
        if (item == null) return new Result({ data: null });
        if (this.#creator) item = this.#creator(item);
        if (!(await this.onValidateDeleteAsync(item))) return new Result({ errors: this.#errors });
        await this.#dbSet.removeAsync(id);
        return new Result({ data: item });
    }

    /**
     * @param {(json: TItem) => TItem} creator
     * @param {DbContext} dbContext 
     * @param {string} name 
     */
    constructor(creator, dbContext, name) {
        super();
        this.#creator = creator;
        this.#dbContext = dbContext;
        this.#dbSet = dbContext.set(name);
        this.#errors = [];
    }

}

export {
    DatabaseCatalogue
}